<a href="https://colab.research.google.com/github/malsagova/StudyRepo26/blob/main/%D0%9C%D0%B0%D0%BB%D1%8C%D1%81%D0%B0%D0%B3%D0%BE%D0%B2%D0%B0_%D0%BF%D1%802.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>



1.   Загружаю библиотеки



In [ ]:
import pandas as pd
!pip install -U autogluon
from autogluon.tabular import TabularDataset, TabularPredictor

2.   Подгружаю данные



In [10]:
data = TabularDataset(r'/content/dispensarization_data_2026.csv')

3. Оцениваю размер данных

In [11]:
print(f"Размер данных: {data.shape}")

Размер данных: (1000, 18)


4. Смотрю первые 5 строк для оценки состояния

In [20]:
print(f"\nПервые 5 строк:")
print(data.head())


Первые 5 строк:
   Возраст  Пол_мужской        ИМТ  Окружность_талии_см  САД_мм_рт_ст  \
0       41            1  24.238237            81.037385    113.191318   
1       46            0  22.883365            80.539167    117.319195   
2       64            0  21.991937            78.423645    121.246690   
3       53            0  19.844568            73.436282    119.746372   
4       53            1  24.284829            82.062243    127.068746   

   ДАД_мм_рт_ст  Пульсовое_давление  Глюкоза_натощак_ммоль_л   HbA1c_%  \
0     62.544325           50.646993                 4.595433  4.648933   
1     62.109786           55.209410                 4.963993  4.000000   
2     71.456074           49.790616                 4.712639  5.034903   
3     71.594664           48.151709                 4.430998  5.080881   
4     70.342105           56.726641                 4.996974  5.044521   

   ЛПНП_ммоль_л  ЛПВП_ммоль_л  Триглицериды_ммоль_л  СКФ_мл_мин  Курение  \
0           NaN        

5. Распределение целевой переменной

In [13]:
print("Распределение 'Доклинический_риск':")
print(data['Доклинический_риск'].value_counts())
print(f"\nДоля единиц: {data['Доклинический_риск'].mean():.2%}")

Распределение 'Доклинический_риск':
Доклинический_риск
0    939
1     61
Name: count, dtype: int64

Доля единиц: 6.10%


6. Делю на выборки

In [14]:
# Обучающая выборка — 80%, тестовая — 20%
train_data = data.sample(frac=0.8, random_state=42)
test_data = data.drop(train_data.index)

print(f"Обучающая выборка: {len(train_data)} строк")
print(f"Тестовая выборка: {len(test_data)} строк")
print(f"\nРаспределение в обучении:")
print(train_data['Доклинический_риск'].value_counts(normalize=True))

Обучающая выборка: 800 строк
Тестовая выборка: 200 строк

Распределение в обучении:
Доклинический_риск
0    0.94875
1    0.05125
Name: proportion, dtype: float64


7. Создаем predictor и запускаю обучение моделей

In [15]:
predictor = TabularPredictor(
    label='Доклинический_риск',
    eval_metric='roc_auc',
    verbosity=2
)

print("Начинаем обучение...")
predictor.fit(train_data)
print("Обучение завершено!")

No path specified. Models will be saved in: "AutogluonModels/ag-20260518_213746"
Verbosity: 2 (Standard Logging)


Начинаем обучение...


=================== System Info ===================
AutoGluon Version:  1.5.0
Python Version:     3.12.13
Operating System:   Linux
Platform Machine:   x86_64
Platform Version:   #1 SMP Thu Apr 30 18:17:14 UTC 2026
CPU Count:          2
Pytorch Version:    2.9.1+cu128
CUDA Version:       CUDA is not available
Memory Avail:       11.07 GB / 12.67 GB (87.4%)
Disk Space Avail:   75.58 GB / 107.72 GB (70.2%)
No presets specified! To achieve strong results with AutoGluon, it is recommended to use the available presets. Defaulting to `'medium'`...
	Recommended Presets (For more details refer to https://auto.gluon.ai/stable/tutorials/tabular/tabular-essentials.html#presets):
	presets='extreme'  : New in v1.5: The state-of-the-art for tabular data. Massively better than 'best' on datasets <100000 samples by using new Tabular Foundation Models (TFMs) meta-learned on https://tabarena.ai: TabPFNv2, TabICL, Mitra, TabDPT, and TabM. Requires a GPU and `pip install autogluon.tabular[tabarena]` to in

[1000]	valid_set's binary_logloss: 0.0291147


	0.9967	 = Validation score   (roc_auc)
	5.73s	 = Training   runtime
	0.01s	 = Validation runtime
Fitting model: LightGBM ...
	Fitting with cpus=1, gpus=0, mem=0.0/11.0 GB
	1.0	 = Validation score   (roc_auc)
	0.54s	 = Training   runtime
	0.01s	 = Validation runtime
Fitting model: RandomForestGini ...
	Fitting with cpus=2, gpus=0


[1000]	valid_set's binary_logloss: 5.8252e-05


	1.0	 = Validation score   (roc_auc)
	1.17s	 = Training   runtime
	0.1s	 = Validation runtime
Fitting model: RandomForestEntr ...
	Fitting with cpus=2, gpus=0
	1.0	 = Validation score   (roc_auc)
	0.95s	 = Training   runtime
	0.1s	 = Validation runtime
Fitting model: CatBoost ...
	Fitting with cpus=1, gpus=0
	1.0	 = Validation score   (roc_auc)
	9.62s	 = Training   runtime
	0.0s	 = Validation runtime
Fitting model: ExtraTreesGini ...
	Fitting with cpus=2, gpus=0
	0.9852	 = Validation score   (roc_auc)
	0.9s	 = Training   runtime
	0.1s	 = Validation runtime
Fitting model: ExtraTreesEntr ...
	Fitting with cpus=2, gpus=0
	0.9868	 = Validation score   (roc_auc)
	0.89s	 = Training   runtime
	0.1s	 = Validation runtime
Fitting model: NeuralNetFastAI ...
	Fitting with cpus=1, gpus=0, mem=0.0/10.9 GB
	0.9482	 = Validation score   (roc_auc)
	6.32s	 = Training   runtime
	0.09s	 = Validation runtime
Fitting model: XGBoost ...
	Fitting with cpus=1, gpus=0
	1.0	 = Validation score   (roc_auc)
	1.05

Обучение завершено!


8. Оценка на тестовой выборке

In [16]:
performance = predictor.evaluate(test_data)

print("\nРезультаты на тестовой выборке:")
for metric, score in performance.items():
    print(f"  {metric}: {score:.4f}")


Результаты на тестовой выборке:
  roc_auc: 0.9806
  accuracy: 0.9800
  balanced_accuracy: 0.9222
  mcc: 0.8852
  f1: 0.8947
  precision: 0.9444
  recall: 0.8500


In [17]:
leaderboard = predictor.leaderboard(test_data, silent=True)
print("\nСравнение всех моделей:")
print(leaderboard[['model', 'score_test', 'pred_time_test', 'fit_time']].head(10))


Сравнение всех моделей:
                 model  score_test  pred_time_test  fit_time
0             CatBoost    0.998889        0.008791  9.623024
1     RandomForestEntr    0.990417        0.111995  0.948744
2     RandomForestGini    0.982917        0.124309  1.169426
3             LightGBM    0.980556        0.018436  0.536288
4  WeightedEnsemble_L2    0.980556        0.021731  0.586469
5        LightGBMLarge    0.978333        0.032953  1.295863
6              XGBoost    0.975000        0.023755  1.048060
7           LightGBMXT    0.970000        0.031149  5.730221
8       ExtraTreesEntr    0.962778        0.121160  0.886250
9       NeuralNetTorch    0.957222        0.013476  9.727536


Модель показала отличные результаты на тестовой выборке (200 пациентов):

ROC-AUC = 0.98 — модель отлично различает пациентов с риском и без него

Accuracy = 98% — правильно предсказано 196 из 200 случаев

F1-score = 0.89 — хороший баланс между точностью и полнотой

Recall = 0.85 — модель выявляет 85% реальных случаев риска

Модель успешно справляется с сильным дисбалансом классов (всего 6% положительных примеров) и готова к практическому применению.

9. Определяем важность признаков

In [18]:
feature_importance = predictor.feature_importance(test_data)

feature_importance_sorted = feature_importance.sort_values('importance', ascending=False)

print("\n Важность признаков для модели:")
print(feature_importance_sorted[['importance', 'stddev']].round(4))

Computing feature importance via permutation shuffling for 17 features using 200 rows with 5 shuffle sets...
	8.3s	= Expected runtime (1.66s per shuffle set)
	1.93s	= Actual runtime (Completed 5 of 5 shuffle sets)



 Важность признаков для модели:
                               importance  stddev
Глюкоза_натощак_ммоль_л            0.1052  0.0227
ИМТ                                0.0546  0.0163
ЛПНП_ммоль_л                       0.0242  0.0057
ССЗ_риск_высокий                   0.0017  0.0023
Физическая_активность_мин_нед      0.0004  0.0010
ЛПВП_ммоль_л                       0.0004  0.0005
Курение                            0.0000  0.0000
Возраст                           -0.0001  0.0010
Статус_глюкозы                    -0.0001  0.0006
СКФ_мл_мин                        -0.0007  0.0009
САД_мм_рт_ст                      -0.0007  0.0024
HbA1c_%                           -0.0012  0.0016
Триглицериды_ммоль_л              -0.0015  0.0010
Окружность_талии_см               -0.0016  0.0019
Пол_мужской                       -0.0022  0.0017
ДАД_мм_рт_ст                      -0.0027  0.0021
Пульсовое_давление                -0.0039  0.0043


Выделены три ключевых фактора доклинического риска:

Глюкоза_натощак_ммоль_л — самый важный признак (важность 0.105)

ИМТ (индекс массы тела) — второй по значимости (важность 0.055)

ЛПНП_ммоль_л (плохой холестерин) — третий значимый фактор (важность 0.024)

Остальные признаки (курение, пол, артериальное давление и др.) показали близкую к нулю важность, что может указывать на их меньшую прогностическую ценность в данной модели.



## Общий краткий вывод по работе

В ходе работы я построила модель для прогнозирования доклинического риска на основе датасета.

**Ключевые результаты:**
- Модель показала **высокое качество** на тестовой выборке: ROC-AUC = 0.98, Accuracy = 98%
- Лучший алгоритм — **CatBoost** (ROC-AUC = 1.0 на валидации)
- **Главные факторы риска**: глюкоза натощак, ИМТ и ЛПНП

**Итог:** Модель успешно справляется с задачей классификации, устойчива к дисбалансу классов и может быть использована для скрининга доклинического риска в медицинской практике.